firedrake/ml_coupling/pytorch/backend.py

```diff
@@ -0,0 +1,83 @@
+from firedrake.function import Function
+from firedrake.vector import Vector
+from firedrake.constant import Constant
+from firedrake.ml_coupling.backend_base import AbstractMLBackend
+
+import firedrake.utils as utils
+
+
+class PytorchBackend(AbstractMLBackend):
+
+    def __init__(self):
+        try:
+            import torch
+            self._backend = torch
+        except ImportError:
+            self._backend = None
+
+    @property
+    def backend(self):
+        if self:
+            return self._backend
+        else:
+            raise ImportError("Error when trying to import PyTorch.")
+
+    def __bool__(self):
+        return self._backend is not None
+
+    @utils.cached_property
+    def custom_operator(self):
+        from firedrake.ml_coupling.pytorch.pytorch_custom_operator import FiredrakeTorchOperator
+        return FiredrakeTorchOperator().apply
```

Comment:
`FiredrakeTorchOperator.apply`? It looks `apply` is a staticmethod.

Reply:
Yes, I am triggering the instantiation on purpose to catch the use of `FiredrakeTorchOperator` when `torch` is not imported, see [here](https://github.com/firedrakeproject/firedrake/blob/1dd095cb5261f9893b9e52efabb518f218690720/firedrake/ml_coupling/pytorch/pytorch_custom_operator.py#L30). I couldn't come up with a way of catching that case at the class level without instantiating the class.

Reply:
Thanks! I will change that.

> And for raising an exception if torch is not imported should this not be done in PytorchBackend.__init__? Surely it is not sensible to create an instance of PytorchBackend if we can't import torch.

Not raising the error eagerly when loading the backend is motivated by having FiredrakeTorchOperator subclassing a class from a package (`torch`) that may not be installed, see [here](https://github.com/firedrakeproject/firedrake/blob/1dd095cb5261f9893b9e52efabb518f218690720/firedrake/ml_coupling/pytorch/pytorch_custom_operator.py#L34).

Reply:
Myself and @JDBetteridge have had a thought about all this import guarding magic. Could we not just have `ml_coupling` as a Firedrake subpackage that is not imported by default?
The pattern for writing a ML-coupled Firedrake thing would then be:
```py
from firedrake import *  # would run even if pytorch not installed
from firedrake.ml.pytorch import *  # would fail if pytorch is not installed

# rest of script...
```

Reply:
I feel like it would make the code much more readable, easier to maintain and generally less "hacky".

Is the 1 line you reference something like `load_backend("pytorch")`? If so then the import could set the `backend` parameter instead.

Also, I don't fully understand the need for us to set a `backend` at all. What would be the issue with not explicitly setting a `backend` and instead just creating `FiredrakeTorchOperators` as required?

Reply:
I am not really convinced it is more readable from a user perspective. No, the line in question is:

```python3
G = torch_operator(...)
```

which does the actually wrapping around a Firedrake operator. The backend class is there for backend-specific operations required for example when mapping tensors from firedrake to the backend and vice-versa. This is for separating concerns between generic operations that apply to any ml backend and backend-specific implementations. This is exactly what we currently do with the adjoint (see [here](https://github.com/firedrakeproject/firedrake/blob/203ec3a32cc7b74e5225c08cc9bbebedf502aef4/firedrake/adjoint/blocks.py#L13)).

Reply:
I don’t understand why it’s necessary to keep track of a backend parameter. I think a regular abstract base class inheritance pattern would suffice. I’d be happy to discuss all of this at the next Firedrake meeting.

Reply:
Sure, we can discuss that at the next meeting.